rust/kernel/sync/condvar.rs

```diff
@@ -0,0 +1,251 @@
+///
+/// The following is an example of using a condvar with a mutex:
+///
+/// ```
+/// use kernel::sync::{CondVar, Mutex};
+/// use kernel::{new_condvar, new_mutex};
+///
+/// #[pin_data]
+/// pub struct Example {
+///     #[pin]
+///     value: Mutex<u32>,
+///
+///     #[pin]
+///     value_changed: CondVar,
+/// }
+///
+/// /// Waits for `e.value` to become `v`.
+/// fn wait_for_vaue(e: &Example, v: u32) {
+///     let mut guard = e.value.lock();
+///     while *guard != v {
+///         e.value_changed.wait_uninterruptible(&mut guard);
+///     }
+/// }
+///
+/// /// Increments `e.value` and notifies all potential waiters.
+/// fn increment(e: &Example) {
+///     *e.value.lock() += 1;
+///     e.value_changed.notify_all();
+/// }
+///
+/// /// Allocates a new boxed `Example`.
+/// fn new_example() -> Result<Pin<Box<Example>>> {
+///     Box::pin_init(pin_init!(Example {
+///         value <- new_mutex!(0, "Example::vlaue"),
+///         value_changed <- new_condvar!("Example::value_changed"),
+///     }))
+/// }
+/// ```
+///
+/// The following is an example of using a condvar with an atomic variable:
+///
+/// ```
+/// use kernel::sync::CondVar;
+/// use core::sync::atomic::{AtomicU32, Ordering};
+///
+/// /// Example.
+/// #[pin_data]
+/// pub struct Example {
+///     value: AtomicU32,
+///
+///     #[pin]
+///     value_changed: CondVar,
+/// }
+///
+/// /// Waits for `e.value` to become `v`.
+/// fn wait_for_vaue(e: &Example, v: u32) {
+///     while e.value.load(Ordering::Acquire) != v {
+///         e.value_changed.wait_nolock_uninterruptible();
+///     }
+/// }
+///
+/// /// Increments `e.value` and notifies all potential waiters.
+/// fn increment(e: &Example) {
+///     e.value.fetch_add(1, Ordering::Release);
+///     e.value_changed.notify_all();
+/// }
+///
+/// /// Allocates a new boxed `Example`.
+/// fn new_example() -> Result<Pin<Box<Example>>> {
+///     Box::pin_init(pin_init!(Example {
+///         value: AtomicU32::new(0),
+///         value_changed <- new_condvar!("Example::value_changed"),
+///     }))
+/// }
+/// ```
+///
+/// [`struct wait_queue_head`]: ../../../include/linux/wait.h
+#[pin_data]
+pub struct CondVar {
+    #[pin]
+    pub(crate) wait_list: Opaque<bindings::wait_queue_head>,
+
+    /// A condvar needs to be pinned because it contains a [`struct list_head`] that is
+    /// self-referential, so it cannot be safely moved once it is initialised.
+    #[pin]
+    _pin: PhantomPinned,
+}
+
+// SAFETY: `CondVar` only uses a `struct wait_queue_head`, which is safe to use on any thread.
+#[allow(clippy::non_send_fields_in_send_ty)]
+unsafe impl Send for CondVar {}
+
+// SAFETY: `CondVar` only uses a `struct wait_queue_head`, which is safe to use on multiple threads
+// concurrently.
+unsafe impl Sync for CondVar {}
+
+impl CondVar {
+    /// Constructs a new condvar initialiser.
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
```

Comment:
I would expect a `new()` method to return a new object. Would it make sense to name this method `new_initializer()`, `create_initializer()`, `new_pin_init()` or something like that?

Reply:
There is no way to use such an initializer manually, at some point you supply it to some function that uses it to initialize the value in some location, e.g.:
```rust
let some_obj = Box::pin_init(SomeObject::new());
// or
let this = Arc::pin_init(pin_init!(Self {
    some_obj <- SomeObject::new(),
}));
```
I think having to write `new_initializer` or similar is going to become cumbersome very fast.

Reply:
Here I am just using the naming convention that Benno and Gary came up with for pin-init.

Being a little pedantic about terminology: `new()` does create a new object, it's just that its type is not `Self` (and you need extra work to get to an instance of `Self`).

One positive of calling it `new` is that we make it clear to users that `CondVar`, `Mutex`, `SpinLock`, etc. have something different about them, that one can't just use it like other types. Using a different name will make it slightly less obvious: users may still look for different ways to construct an instance of such types.